insecure/corruptible/factory_test.go

```diff
@@ -0,0 +1,225 @@
+package corruptible
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	testifymock "github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+
+	"github.com/onflow/flow-go/insecure"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/libp2p/message"
+	"github.com/onflow/flow-go/network"
+	"github.com/onflow/flow-go/network/codec/cbor"
+	"github.com/onflow/flow-go/network/mocknetwork"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+func TestRegisterAdapter(t *testing.T) {
+	f := NewCorruptibleConduitFactory(flow.BftTestnet, unittest.IdentifierFixture(), cbor.NewCodec())
+
+	adapter := &mocknetwork.Adapter{}
+
+	// registering adapter must go successful
+	require.NoError(t, f.RegisterAdapter(adapter))
+
+	// second attempt on registering adapter must fail
+	require.Error(t, f.RegisterAdapter(adapter))
+}
+
+// TestNewConduit_HappyPath checks when factory has an adapter registered, it can successfully
+// create conduits.
+func TestNewConduit_HappyPath(t *testing.T) {
+	f := NewCorruptibleConduitFactory(flow.BftTestnet, unittest.IdentifierFixture(), cbor.NewCodec())
+	channel := network.Channel("test-channel")
+
+	adapter := &mocknetwork.Adapter{}
+
+	require.NoError(t, f.RegisterAdapter(adapter))
+
+	c, err := f.NewConduit(context.Background(), channel)
+	require.NoError(t, err)
+	require.NotNil(t, c)
+}
+
+// TestNewConduit_MissingAdapter checks when factory does not have an adapter registered,
+// any attempts on creating a conduit fails with an error.
+func TestNewConduit_MissingAdapter(t *testing.T) {
+	f := NewCorruptibleConduitFactory(flow.BftTestnet, unittest.IdentifierFixture(), cbor.NewCodec())
+	channel := network.Channel("test-channel")
+
+	c, err := f.NewConduit(context.Background(), channel)
+	require.Error(t, err)
+	require.Nil(t, c)
+}
+
+// TestFactoryHandleIncomingEvent_AttackerObserve evaluates that the incoming messages to the conduit factory are routed to the
+// registered attacker if one exists.
+func TestFactoryHandleIncomingEvent_AttackerObserve(t *testing.T) {
```

Comment:
I don’t see where `Observe()` is called for the attacker. Can there be more explanation how we set up simulating routing messages to the attacker?

Reply:
Added comment in https://github.com/onflow/flow-go/pull/2033/commits/55f291604dbfde4ad3ec9fd9a53f397a2926573a.